cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+
+We start with a `term`, which is just a schema for data:
+
+```lisp
+    (decl Term (A B C u32 u32) T)
+```
+
+A term can have:
+
+1. A single internal extractor body, via a toplevel `(extractor ...)`
+   form, OR
+   
+2. A single external extractor binding, via an `(extern extractor
+   ...)` form, defined below; AND
+   
+3. One or more `(rule (Term ...) ...)` toplevel forms, which together
+   make up an internal constructor definition, OR
+   
+4. A single external constructor binding, via an `(extern constructor
+   ...)` form, defined below.
+
+## ISLE to Rust
+
+Now that we have described the core ISLE language, we will document
+how it interacts with Rust code. We consider these interactions to be
+semantically as important as the core language: they are not
+implementation details, but rather, a well-defined interface by which
+ISLE can interface with the outside world (an "FFI" of sorts).
+
+### Mapping to Rust: Constructors, Functions, and Control Flow
+
+ISLE was designed to have a simple, easy-to-understand mapping from
+its language semantics to Rust semantics. This means that the
+execution of ISLE rewriting has a well-defined implementation in
+Rust. The basic principles are:
+
+1. Every constructor with rules in ISLE becomes a Rust function. The
+   constructor arguments are the Rust function arguments. The
+   constructor's "return value" is the Rust function's return value
+   (usually wrapped in an `Option` because pattern coverage can be
+   incomplete).
+
+2. One rewrite step is one Rust function call.
+
+3. Rewriting is thus eager, and reified through ordinary Rust control
+   flow. The code that embeds the ISLE generated code will kick off
+   execution by calling a top-level "driver" constructor; this will
+   invoke constructors in its rewrite expression, making other Rust
+   function calls, until eventually a value is returned.
+   
+4. This design means that "intermediate terms" -- constructed terms
+   that are then further rewritten -- are never actually built as
+   in-memory data-structures. Rather, they exist only as ephemeral
+   stack-frames while the corresponding Rust function executes. This
+   means that there is very little or no performance penalty to
+   factoring code into many sub-rules (subject only to function-call
+   overhead and/or the effectiveness of the Rust inliner).
+   
+5. Backtracking -- attempting to match rules, and backing up to follow
+   a different path when a match fails -- is entirely internal to the
+   generated Rust function for a constructor. Once we are rewriting a
+   constructor and its subterms, we have committed to that
+   constructor. Internally to the rewrite-evaluation of one
+   constructor, it evaluates left-hand side patterns, finding one that
+   matches, and then commits to that and starts to invoke constructors
+   to build the rewritten right-hand side.
+   
+   Said another way, the principle is that left-hand sides are
```

Comment:
This is a bit confusing since before in the general parts "constructor" was also used for LHS. I'm also not seeing yet how this works with the heuristic of precedence for more-specific rules (though maybe that's coming)

Reply:
Updated language here to talk about "terms" instead, thanks for pointing this out!